Consul: Update monitor/telemetry/telegraf for v1.21 and v1.22 (#1504)
hashicorp/web-unified-docs: content/consul/v1.21.x/content/docs/monitor/telemetry/telegraf.mdx

This page describes the process to set up Telegraf to monitor Consul datacenter telemetry.

> Consul makes a range of metrics in various formats available so operators can measure the health and stability of a datacenter, and diagnose or predict potential issues.

Reviewer: ```suggestion
## Introduction

Consul makes a range of metrics in various formats available so operators can measure the health and stability of a datacenter, as well as diagnose and predict potential issues.
```

> In this example you are going to use the [telegraf_plugin][] in conjunction with the StatsD protocol supported by Consul. For the full list of metrics available with Consul, refer to the [telemetry documentation](/consul/docs/reference/agent/telemetry).

Reviewer: ```suggestion
One monitoring solution is to use the [telegraf_plugin][] in conjunction with the StatsD protocol supported by Consul. You can also use this data with Grafana to organize and query the data you collect.

For the full list of Consul agent metrics, refer to the [telemetry documentation](/consul/docs/reference/agent/telemetry).
```

This page describes the process to set up Telegraf to monitor Consul datacenter telemetry.

## Introduction

Consul makes a range of metrics in various formats available so operators can measure the health and stability of a datacenter, and diagnose or predict potential issues.

> One monitoring solution is to use the [telegraf_plugin][] in conjunction with the StatsD protocol supported by Consul. You can also use this data with Grafana to organize and query the data you collect.

Reviewer: ```suggestion
One monitoring solution is to use the [Telegraf Consul plugin](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/consul) with the StatsD protocol supported by Consul. You can also use Grafana to organize and query the data you collect.
```
Can you update the outdated hyperlink formatting? So instead of the double square brackers `[][]` it's a markdown hyperlink `[]()`.

For the full list of Consul agent metrics, refer to the [telemetry documentation](/consul/docs/reference/agent/telemetry).

## Workflow

1. [Configure Telegraf to collect StatsD and host level metrics](#configure-telegraf)
1. [Configure Consul to send metrics to Telegraf](#configure-consul)
1. [Review Consul metrics](#review-consul-metrics)

## Configure Telegraf

> Telegraf acts as a StatsD agent and can collect additional metrics about the hosts where Consul agents are running. Telegraf itself ships with a wide range of [input plugins][telegraf-input-plugins] to collect data from lots of sources for this purpose.

Reviewer: ```suggestion
Telegraf acts as a StatsD agent and can collect additional metrics about the hosts where Consul agents are running. 
```

> You are going to enable some of the most common input plugins to monitor CPU, memory, disk I/O, networking, and process status, since these are useful for debugging Consul datacenter issues. Here is an example `telegraf.conf` file that you can use as a starting point:

Reviewer: ```suggestion
Telegraf includes input plugins to collect data such as CPU usage, memory usage, disk I/O, networking, and process status. The following example uses a `telegraf.conf` file configured to debug common Consul datacenter issues.
```

The next section of `telegraf/conf` sets up a StatsD listener on UDP port 8125 with instructions to calculate percentile metrics and to parse DogStatsD-compatible tags. Consul uses this data to report telemetry stats. The full reference to all the available StatsD-related options in Telegraf is [here][telegraf-statsd-input].

The next configuration sections are used to configure inputs for things like CPU, memory, network I/O, and disk I/O. It is important to make sure the `interfaces` list in `inputs.net` matches the system interface names. Most Linux systems use names like `eth0` or `enp0s0`, but you can choose any valid interface name from your system. The list also supports glob patterns, for example `eth*` will match all interfaces starting with `eth`.

Another useful input plugin is the [procstat Telegraf plugin][telegraf-procstat-input], which reports metrics for a process according to a given pattern. In this case, you are using it to monitor the Consul agent process itself.

> Telegraf even includes a [plugin that monitors the health checks associated with the Consul agent][telegraf-consul-input], using the Consul API to query the data.

Reviewer: ```suggestion
Finally, the configuration includes a [plugin that monitors the health checks associated with the Consul agent][telegraf-consul-input] by using the Consul API to query the data.
```

> 

Reviewer: Include one more sentence here about applying the configuration to your Telegraf instance.

[[inputs.system]]
  # no configuration

[[inputs.procstat]]
  pattern = "(consul)"

[[inputs.consul]]
  address = "localhost:8500"
  scheme = "http"
```

</CodeBlockConfig>

The `telegraf.conf` file starts with global options - you set the default collection interval to 10 seconds and ask Telegraf to include a `host` tag in each metric.

Telegraf also allows you to set additional tags on the metrics that pass through it. In this case, you are adding tags for the server role `consul-server` and datacenter `us-east-1`. You can further use these tags in Grafana to filter queries.

> The next config section sets up a StatsD listener on UDP port 8125, with instructions to calculate percentile metrics and to parse DogStatsD-compatible tags. Consul will use this to report telemetry stats. The full reference to all the available StatsD-related options in Telegraf is [here][telegraf-statsd-input].

Reviewer: ```suggestion
The next section of `telegraf/conf` sets up a StatsD listener on UDP port 8125 with instructions to calculate percentile metrics and to parse DogStatsD-compatible tags. Consul uses this data to report telemetry stats. The full reference to all the available StatsD-related options in Telegraf is [here][telegraf-statsd-input].
```

</CodeTabs>

> Note that the configuration specifies DogStatsD format instead of plain StatsD, which tells Consul to send [tags][tagging] with each metric. Tags can be used by Grafana to filter data on your dashboards (for example, displaying only the data for which `role=consul-server`). Telegraf is compatible with the DogStatsD format and allows you to add your own tags too.

Reviewer: ```suggestion
The configuration specifies DogStatsD format instead of plain StatsD. As a result, Consul sends [tags][tagging] with each metric. You can use Grafana to filter data on your dashboards according to these tags. For example, you can display server agent data by filtering for `role=consul-server`. Telegraf is compatible with the DogStatsD format, and allows you to add your own tags too.
```

Here is an example Grafana dashboard:

![Grafana Consul Datacenter](/img/consul-grafana-screenshot.png 'Grafana Dashboard')

Some of the important metrics to monitor include:

- [Memory usage metrics](#memory-usage-metrics)
- [File descriptor metrics](#file-descriptor-metrics)
- [CPU usage metrics](#cpu-usage-metrics)
- [Network activity metrics](#network-activity-metrics)
- [Disk activity metrics](#disk-activity-metrics)

### Memory usage metrics

| Metric Name         | Description                                                    |
| :------------------ | :------------------------------------------------------------- |
| `mem.total`         | Total amount of physical memory (RAM) available on the server. |
| `mem.used_percent`  | Percentage of physical memory in use.                          |
| `swap.used_percent` | Percentage of swap space in use.                               |

**Why they're important:** Consul keeps all of its data in memory. If Consul consumes all available memory, it will crash. You should also monitor total available RAM to make sure some RAM is available for other processes, and swap usage should remain at 0% for best performance.

> **What to look for:** If `mem.used_percent` is over 90%, or if `swap.used_percent` is greater than 0.

Reviewer: ```suggestion
**When to take action:** If `mem.used_percent` is over 90%, or if `swap.used_percent` is greater than 0.
```

### File descriptor metrics

| Metric Name                | Description                                                         |
| :------------------------- | :------------------------------------------------------------------ |
| `linux_sysctl_fs.file-nr`  | Number of file handles being used across all processes on the host. |
| `linux_sysctl_fs.file-max` | Total number of available file handles.                             |

> **Why it's important:** Practically anything Consul does -- receiving a connection from another host, sending data between servers, writing snapshots to disk -- requires a file descriptor handle. If Consul runs out of handles, it will stop accepting connections. Check [the Consul FAQ][consul_faq_fds] for more details.

Reviewer: ```suggestion
**Why they are important:** Practically anything Consul does, from receiving a connection from another host to sending data between servers or writing snapshots to disk, requires a file descriptor handle. If Consul runs out of handles, it will stop accepting connections. Refer to [the Consul FAQ][consul_faq_fds] for more details.
```

By default, process and kernel limits are fairly conservative. You will want to increase these beyond the defaults.

> **What to look for:** If `file-nr` exceeds 80% of `file-max`.

Reviewer: ```suggestion
**When to take action:** If `file-nr` exceeds 80% of `file-max`.
```

### CPU usage metrics

| Metric Name      | Description                                                      |
| :--------------- | :--------------------------------------------------------------- |

> | `cpu.user_cpu`   | Percentage of CPU being used by user processes (such as Consul). |

Reviewer: ```suggestion
| `cpu.user_cpu`   | Percentage of CPU being used by user processes, including Consul. |
```

| `cpu.iowait_cpu` | Percentage of CPU time spent waiting for I/O tasks to complete.  |

**Why they're important:** Consul is not particularly demanding of CPU time, but a spike in CPU usage might indicate too many operations taking place at once, and `iowait_cpu` is critical -- it means Consul is waiting for data to be written to disk, a sign that Raft might be writing snapshots to disk too often.

> **What to look for:** if `cpu.iowait_cpu` greater than 10%.

Reviewer: ```suggestion
**When to take action:** If `cpu.iowait_cpu` is greater than 10%.
```

**What to look for:** Sudden large changes to the `net` metrics (greater than 50% deviation from baseline).

**NOTE:** The `net` metrics are counters, so in order to calculate rates (such as bytes/second), you will need to apply a function such as [non_negative_difference][].

### Disk activity metrics

| Metric Name          | Description                         |
| :------------------- | :---------------------------------- |
| `diskio.read_bytes`  | Bytes read from each block device.  |
| `diskio.write_bytes` | Bytes written to each block device. |

**Why they're important:** If the Consul host is writing a lot of data to disk, such as under high volume workloads, there may be frequent major I/O spikes during leader elections. This is because under heavy load, Consul is checkpointing Raft snapshots to disk frequently.

It may also be caused by Consul having debug/trace logging enabled in production, which can impact performance.

Too much disk I/O can cause the rest of the system to slow down or become unavailable, as the kernel spends all its time waiting for I/O to complete.

> **What to look for:** Sudden large changes to the `diskio` metrics (greater than 50% deviation from baseline, or more than 3 standard deviations from baseline).

Reviewer: ```suggestion
**When to take action:** You experience sudden large changes to the `diskio` metrics that are greater than 50% deviation or more than 3 standard deviations from baseline.
```

## Next steps

To read further about telemetry in Consul, check the [Consul Agent Telemetry](/consul/docs/monitor/telemetry/agent) and [Consul Dataplane Telemetry](/consul/docs/monitor/telemetry/dataplane) pages.

> To learn more about Consul monitoring, alerting and logging, check out the [Consul Monitoring](/consul/docs/monitor) page.

Reviewer: ```suggestion
To learn more about monitoring, alerting, and logging data generated by Consul agents, refer to [Consul Monitoring](/consul/docs/monitor).
```
